### ISSUE TITLE
New LTR/RTL styles on 3.4.0 override custom CSS rules

### ISSUE THREAD

**Issue by FdezRomero** [ionic-team/ionic#12053]
**Ionic version:**  (check one with "x")
[ ] **1.x** (For Ionic 1.x issues, please use https://github.com/ionic-team/ionic-v1)
[ ] **2.x**
[x] **3.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
The new CSS rules introduced in 3.4.0 for LTR/RTL are prefixed with the attribute selector `[dir="ltr"]` or `[dir="rtl"]` (from the root `<html>` tag), but these rules always having a higher specificity than the ones you would normally write inside a page break the custom styles of our app.

The only solutions we came up with are adding specificity to our current selectors (would require to top 4 classes/attributes), using `!important` on every conflicting rule (not recommended for so many reasons) or prefixing the page selectors in every SCSS file, like `[dir] page-home {...}`.

**Expected behavior:**
LTR/RTL selectors to use a lower specificity so we can override Ionic rules easily and cleanly, as we have been doing. Splitting problematic LTR/RTL joint rules (`[dir="ltr"] .some-class, [dir="rtl"] .some-class`) in two would halve specificity to `0, 0, 2, 0`.

**Steps to reproduce:**
1. Add a `<button ion-button class="my-button"></button>` to a page.
2. Try to style the button from SCSS, like `page-name { .my-button { padding: 16px; margin: 0; } }`.
3. Your rules are overwritten by the `[dir="ltr"] .button-(mode), [dir="rtl"] .button-(mode)` selector.
4. This happens with almost every component: `ion-card`, `ion-card-content`, `ion-chip`, etc.

**Related code:**
N/A

**Other information:**
N/A

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
global packages:

    @ionic/cli-utils : 1.4.0
    Cordova CLI      : 7.0.1
    Ionic CLI        : 3.4.0

local packages:

    @ionic/app-scripts              : 1.3.7
    @ionic/cli-plugin-cordova       : 1.4.0
    @ionic/cli-plugin-ionic-angular : 1.3.1
    Cordova Platforms               : android 6.2.3 ios 4.4.0
    Ionic Framework                 : ionic-angular 3.4.0

System:

    Node       : v6.10.3
    OS         : macOS Sierra
    Xcode      : Xcode 8.3.3 Build version 8E3004b
    ios-deploy : 1.9.1
    ios-sim    : 5.1.0
    npm        : 5.0.3
```

**Comment by brandyscarney** [comment]
@FdezRomero Thanks for the issue! Are you only using one direction or does your app require both? If you only need one direction can you set the following variable in your Sass and let me know if that works for you:

```
$app-direction: ltr;
```

**Comment by AmitMY** [comment]
That is fair.
What I suggest doing is either setting a strict direction like @brandyscarney suggested, or use our new sass mixins to handle multi directional apps.
See our newly released [blog post](http://blog.ionic.io/ionic-and-rtl/) in the matter.

**Comment by FdezRomero** [comment]
Thanks @brandyscarney @AmitMY! I confirm that adding `$app-direction: ltr;` in `variables.scss` removes the LTR/RTL selectors and everything works as usual. 🎉

It would be a good idea to add this to both the changelog and release info, as it may be considered a breaking change.

**Comment by brandyscarney** [comment]
@FdezRomero I am releasing v3.4.1 which will set it to `ltr` by default. It should be out any minute.

**Comment by FdezRomero** [comment]
Thanks for the heads up @brandyscarney! You are doing an amazing job, you really are 😎👌

**Comment by brandyscarney** [comment]
Thank you! So I actually just released a `3.4.2` after finding an issue with changing the default `$app-direction` - so it will be `ltr` by default, can be set to `rtl` for rtl only or `multi` for both. We'll have the docs updated to reflect this soon. Thanks for bringing this to our attention!

**Comment by nottinhill** [comment]
This has cost us 2 weeks of debugging time, approx. $5.000,00.

**Comment by AmitMY** [comment]
@nottinhill Next time:
- **Step 1**: Read the CHANGELOG - If you update to a new version, you have to do a bunch of stuff, and be aware of breaking changes if any.
- **Step 2**: Always keep up to date! if you were on 3.4.0 for 2 weeks, and have a bug, you should have updated at least to the version that came 1 day after it, which is a patch version 3.4.2
- **Step 3**: Have a core issue you didn't have before? No worries! Look in the github issues, and you would find this one. Alternatively, create an issue, and I will close it in just a few hours as a duplicate of this issue, and refer you here. [Here ](https://ionicframework.com/support) are the ways you can get support.

Setting that "cost" is really irrelevant, as there are many things you could have done to solve this in 3-8 hours, tops, depending on timezone.

^^I read this message just now, and it sounds a bit aggressive. That is not my intent :), I just want to inform you of what to do next times when something as major as this happens.

**Comment by supryin** [comment]
@AmitMY 
would you kindly point me to a section of the changelog, where it mentions that custom css gets overwritten and the fix of that bug? 
In a complex development environment it is not trivial to understand that this is the issue.

**Comment by AmitMY** [comment]
I can't point you in 3.4.0, as there was no announcement of a breaking change. But, it is just step 1. Steps 2 and 3 are still valid for this case, and each could have got your issue fixed immediately.

This was a non-desired breaking change, and we ran the tests inside the app multiple times, and there was a nightly released that was requested for maintainers to see if it breaks their app, and no one said anything, it was assumed that the version is solid.
However, If you do read the CHANGELOG, 1 day after 3.4.0 there was a patch released [3.4.1](https://github.com/ionic-team/ionic/blob/master/CHANGELOG.md#341-2017-06-16) which fixed this.

I am the one who created this breaking change, and really, didn't think through all of the cases, as v3.4.0 includes a massive amount of changes to the styling (at least 50% of the css selectors in this project), and the tests normally cover style issues, A [nice read](http://blog.ionic.io/snapshot-ionic-frameworks-fail-safe/) about the tests we ran.

I don't see a need to apologize. You have a bug on an open source project? It is your duty (as a user of an open source project) to create an issue, like this issue creator did. Plus, you can always downgrade to a previous version and life would be dandy. (even if you don't know what causes the issue: "v3.4.0 styles are not the same as v3.3.0" is all you have to say for people to get what is the problem or at least investigate.

Note that I do not speak for the employees of Ionic, I am just a guy giving my free time to perfect this awesome framework.

**Comment by nottinhill** [comment]
Thanks for your thoughts. However I must say, since we started using Ionic 2 Beta we burned through a lot of resources and money, nerves, tears and sweats. I am in the business of building reliable infrastructure which seems in the case of ngx and ionic not to be possible without a dedicated DevOps Team RSS-ing onto Ionic's changelog stream 24/7. We never experienced anything as buggy and ever-changing like Ionic and I am really looking forward to future projects ∈ \ Ionic.

**Comment by AmitMY** [comment]
I am a one man team, and I manage to get a grip on the changelog on all of the packages I star on github using [sibbell](https://about.sibbell.com/).
I do wanna correct you on one thing, there is no "changelog stream", it is more like a per-release change, so more like a blog thing.

Glad to help clarify everything.
It is your choice what to use, and thanks for using Ionic at least for now.

**Comment by mhartington** [comment]
Locking this issue as the conversation here is not productive.  If there are any issues, please be sure to update the the latest release. 3.4.0 had a unintentional breaking change, we addressed it by releasing 3.4.2 and notifying people of the release and what it fixes in the changelog.
